### PR TITLE
require kryo registration in tests

### PR DESCRIPTION
**Pull request by ryan-williams**

### CODE REVIEW

**Comment by ryan-williams** [comment]
jenkins, retest this please

File: adam-core/src/test/scala/org/bdgenomics/adam/util/ADAMFunSuite.scala [hunk 1]
```diff
@@ -31,7 +31,9 @@ trait ADAMFunSuite extends SparkFunSuite {
   override val properties: Map[String, String] = Map(
     ("spark.serializer", "org.apache.spark.serializer.KryoSerializer"),
     ("spark.kryo.registrator", "org.bdgenomics.adam.serialization.ADAMKryoRegistrator"),
-    ("spark.kryo.referenceTracking", "true"))
+    ("spark.kryo.referenceTracking", "true"),
+    "spark.kryo.registrationRequired" -> "true"
```
Comment (fnothaft) [inline adam-core/src/test/scala/org/bdgenomics/adam/util/ADAMFunSuite.scala]: Could we have this in tuple form, or convert the other entries to map style? I'd just like all to be consistent.

Comment (ryan-williams) [inline adam-core/src/test/scala/org/bdgenomics/adam/util/ADAMFunSuite.scala]: ah, woops, done!

**Comment by ryan-williams** [comment]
Jenkins, retest this please

**Comment by ryan-williams** [comment]
Last build failure mentions c24aae5619621fe18acc46e7a870301f94e130e9 but seems to check out and test 394bc9ca51e1b910613e4e88680b6e9ae1cbe58b, which I can't even find in my local checkout of ADAM, so I'm not sure what happened there but am hoping a retest will pick up my HEAD commit here, https://github.com/bigdatagenomics/adam/pull/1020/commits/c24aae5619621fe18acc46e7a870301f94e130e9

**Comment by ryan-williams** [comment]
OK, that picked up my last commit, and apparently there's more that I need to register.

What I don't understand is that only two builds in the matrix have been failing this whole time, and I can't reproduce those failures locally.

Jenkins runs:

```
mvn clean package -Dhadoop.version=2.6.0 -Dspark.version=1.5.2
```

and the FlattenSuite in adam-cli fails.

When I run locally:

```
mvn clean test -Dsuites='*FlattenSuite*' -pl adam-cli -Dhadoop.version=2.6.0 -Dspark.version=1.5.2
```

the test passes.

Also only the Spark 1.5.2, Hadoop 2.6.0 builds fail.

**Comment by fnothaft** [comment]
> Also only the Spark 1.5.2, Hadoop 2.6.0 builds fail.

We only run all of the other build configurations if the Spark 1.5.2/Hadoop 2.6.0 builds pass. If they fail, we don't run the other builds.

**Comment by heuermh** [comment]
> What I don't understand is that only two builds in the matrix have been failing this whole time

They are set to be the default build configuration, in that if they fail, the rest are not run.  Unfortunately the Jenkins matrix chart shows green for the rest of the build configurations, even though those were not run.

```bash
$ HADOOP_VERSION=2.6.0 SPARK_VERSION=1.5.2 ./scripts/jenkins-test
```
should replicate what you see in Jenkins.

**Comment by ryan-williams** [comment]
Thanks for the info about the 1.5.2/2.6.0 builds!

Jenkins still seems to not pick up my commit after I force-push. atm my HEAD commit is b527f1887de27b66dc17df5c0ab02def84f71a15; here's what the last Jenkins run shows:

```
originally caused by:
 GitHub pull request #1020 of commit b527f1887de27b66dc17df5c0ab02def84f71a15 automatically merged.
[EnvInject] - Loading node environment variables.
Building remotely on amp-jenkins-worker-04 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb/HADOOP_VERSION/2.6.0/SCALAVER/2.10/SPARK_VERSION/1.5.2/label/centos
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10
Fetching upstream changes from https://github.com/bigdatagenomics/adam.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/*
Checking out Revision c637cf7fe857fdd53e6c25513cb059723e4e7c11 (origin/pr/1020/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f c637cf7fe857fdd53e6c25513cb059723e4e7c11
```

It prints b527f1887de27b66dc17df5c0ab02def84f71a15 at first, but then checks out c637cf7fe857fdd53e6c25513cb059723e4e7c11, which again I can't find anywhere.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
I ran 
```
HADOOP_VERSION=2.6.0 SPARK_VERSION=1.5.2 ./scripts/jenkins-test
```
and the failing test here (`FlattenSuite`) passed; still failing via Jenkins 😢

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
This looks like the usual "commit not found" issue that happens after a force-push; however, I specifically didn't force-push the commit that triggered this build, to try to see if that simplified my iterations here. ❓ 

Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by heuermh** [comment]
@fnothaft @massie @shaneknapp It is possible there is something else at play here?  I had a similar amount of trouble with Jenkins builds on #979.  I don't think it is right to force dealing with all these failed builds on external collaborators.  Not everyone has as much patience as @ryan-williams (and thank you for it!).

What do folks think about dropping Jenkins for pull request builds and using Travis instead?

**Comment by shaneknapp** [comment]
TBH, without any real background in this right now, i'd be really hesitant to ditch jenkins for an entirely new build system setup.  i'll talk to @fnothaft right now and get his input/feelings.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
@heuermh While there are known issues with Jenkins' git support and force pushes, we don't have the build engineering bandwidth to move this build to another system while still supporting all other builds on Jenkins. I haven't been following this pull request in any detail so I can't comment as to whether the build issues are Jenkins related or not other than the Jenkins picking the wrong git commit, which as mentioned, is a known issue.

**Comment by ryan-williams** [comment]
One thing that seems to be happening here that I've not previously observed is Jenkins not finding commits even without me force-pushing.

Since https://github.com/bigdatagenomics/adam/pull/1020/commits/fc60199492ebebd9e944295d82d811495b12371d, I've not force-pushed any commits, and at each regularly-pushed commit since, there's been a failed Jenkins run that only picked up the commit on a retest:
- https://github.com/bigdatagenomics/adam/pull/1020/commits/2ef52bb9f20975ebd53f43a4c0e68c784a1057b1
  - [1193](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1193/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): "Could not checkout".
  - [1194](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1194/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): retest picked up commit.
- https://github.com/bigdatagenomics/adam/pull/1020/commits/92c87a766e9d5e76a376b8675c3a2a2b1f1c726b
  - [1195](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1195/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): "Could not checkout".
  - [1196](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1196/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): retest picked up commit.
- https://github.com/bigdatagenomics/adam/pull/1020/commits/34932cdc7044648cb6e54ed84d7a544c0ef5ed3f
  - [1197](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1197/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): started to run, but the test failure observed implies that https://github.com/bigdatagenomics/adam/pull/1020/commits/34932cdc7044648cb6e54ed84d7a544c0ef5ed3f was not included:
    ```
    Note: To register this class use: kryo.register(org.apache.avro.Schema$LockableArrayList.class);
    ```

  - [1198](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1198/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console): retest picked up https://github.com/bigdatagenomics/adam/pull/1020/commits/34932cdc7044648cb6e54ed84d7a544c0ef5ed3f and showed a different test failure (which is what 1197 should have done).

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Seeing the same thing again: [1199](https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/1199/HADOOP_VERSION=2.6.0,SCALAVER=2.10,SPARK_VERSION=1.5.2,label=centos/console) couldn't find the commit https://github.com/bigdatagenomics/adam/pull/1020/commits/edd2152b8a803fc85b98809a9d8327cfdb181a06 even though it was not force-pushed.

Has this behavior been observed to happen on PRs that have _ever_ involved force-pushed commits?

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by shaneknapp** [comment]
the "could not checkout" errors are caused by github not returning a 200 on
the http connection for the fetch.  i have put much investigation in to
this, and have ruled out anything on our end (meaning it's not excess
network traffic, github rate limiting, etc).

usually a retrigger will get the fetch to actually catch, but i will set up
a local git mirror for the repo on all of the workers.  this was affecting
spark builds really badly, and some databricks folks and i hammered out the
setup and it's worked really well for them.

i'll get the local mirrors set up tomorrow.  it won't take long, or
interrupt builds.

shane

**Comment by ryan-williams** [comment]
Thanks @shaneknapp! Weird that Github is reliably returning not-200 for these fetches for non-force-pushed commits…

**Comment by fnothaft** [comment]
> i'll get the local mirrors set up tomorrow.  it won't take long, or
interrupt builds.

Awesome! Thanks @shaneknapp!

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by shaneknapp** [comment]
yeah, it is weird.  and not reproduceable on the command line (trust me, i
tried).  it's also random enough that it's almost impossible to catch when
it happens, and there's enough traffic that tcpdum/wireshark would be
really hard to set up.  it's most likely something horribly broken in the
interactions between the pull request builder plugin and/or the jenkins git
plugins.

this was my tire fire for a number of months.  :)

**Comment by ryan-williams** [comment]
Gotcha! I seem to be reproducing it pretty reliably here, lmk if you want to try to catch it the next time around, which surely is about to happen given the commit I just pushed…

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by fnothaft** [comment]
Yahtzee!

**Comment by ryan-williams** [comment]
Jenkins, retest this please.

**Comment by ryan-williams** [comment]
(Squashed commits)

**Comment by fnothaft** [comment]
Thanks @ryan-williams for the stellar change and for the patience with @AmplabJenkins! I've just merged this.

**Pull request merged**